l3_dump.py

```diff
@@ -9,40 +9,198 @@
+    # does not spit that out.
+    # --------------------------------------------------------------------------
+    hexdump_pat = re.compile(r'(\s*)(0x)?(?P<offset_hex>[0-9a-fA-F]+)(.*)')
+    for line in input_str.split('\n'):
+        # Skip any leader lines not matching required offset-string lines.
+        line_match = hexdump_pat.match(line)
+        if line_match is None:
+            continue
+
+        _offset = int(line_match.group('offset_hex'), 16)
+        break
+
+    return _offset
+
+
+# #############################################################################
+def parse_rodata_string_offsets(program_bin:str) -> dict:
+    """
+    Parse output from `readelf` to extract start offset of embedded strings.
+    The expected convention is that the strings are null-terminated in the
+    hexdump output.
+
+    We are parsing the following sample hexdump output on Linux:
+
+String dump of section '.rodata':
+  [     8]  /tmp/l3.c-small-unit-test.dat
+  [    26]  Simple-log-msg-Args(1,2)
+  [    3f]  Simple-log-msg-Args(3,4)
+
+    Build a dictionary from 'int-offset' -> string; i.e.
+        [ 8] : "/tmp/l3.c-small-unit-test.dat"
+        [38] : "Simple-log-msg-Args(1,2)"
+
+    ... and so on.
+    """
+    with subprocess.Popen([READELF_BIN, READELF_STRDUMP_ARG,
+                           READELF_DATA_SECTION,
+                           program_bin],
+                          stdout=subprocess.PIPE,
+                          stderr=subprocess.PIPE, text=True) as rodata_hexdump:
+        _stdout, _stderr = rodata_hexdump.communicate()
+
+    # pylint: disable-next=unused-variable
+    (_strings, nlines) = parse_string_offsets(_stdout)
+    return _strings
+
+# #############################################################################
+def parse_string_offsets(input_str:str) -> (dict,int):
+    """
+    Parse an input string that is stdout from `readelf`. Extract start offset
+    of embedded strings. Return a dictionary mapping {offset: string}.
+
+    The expected convention is that the strings are null-terminated in the
+    hexdump output. We only want to parse lines that list string offsets, like:
+
+  [     8]  /tmp/l3.c-small-unit-test.dat
+  [    26]  Simple-log-msg-Args(1,2)
+  [    3f]  Simple-log-msg-Args(3,4)
+
+    Returns: Dictionary mapping {offset: string} and # of valid-lines parsed.
+    """
+    _strings = {}
+
+    # --------------------------------------------------------------------------
+    # Compile a r.e. with named capture-groups to extract offset / msg fields.
+    # Allow for white-spaces in diff parts of the line around '[ ]' and in
+    # the front, and between phrases. This is designed to cope liberally with
+    # `readelf` output that may vary slightly on diff platforms; e.g. Mac/OSX.
+    # --------------------------------------------------------------------------
+    off_string_pat = re.compile(r'(\s*)\[\s*(?P<offset_hex>[0-9a-fA-F]+)(\s*)\](\s+)(?P<msg>.*)')
```

Comment:
Inline parsing done in old code using multiple `split()`s etc. is now replaced by this reliable use of r.e. with named capture-groups.